### PR TITLE
feat: add `InlayHintProvider` for local variable types 

### PR DESCRIPTION
**Pull request by nicdard**
Adding objects to send InlayHints as described in vscode api and LSP 3.17 (upcoming)

I have left out [inlayHintLabelPart](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#inlayHintLabelPart), which we can consider later on to allow also navigation from the tooltip.

See: https://github.com/flix/flix/issues/3383

### CODE REVIEW

**Comment by magnus-madsen** [comment]
Do you want this reviewed and merge first? Or do you want to wait until you also have the Provider?

**Comment by nicdard** [comment]
I think you can already add your comments, since those classes are unlekely to be changed, but as you prefer :)

I will soon add the provider also, I can append it here if you want to do it in one only PR

**Comment by magnus-madsen** [comment]
Lets append it here and then I will look at it together :)

**Comment by nicdard** [comment]
> Lets append it here and then I will look at it together :)

I will change this to draft for now, so you will know when it will be ready afterwards

**Comment by nicdard** [comment]
@magnus-madsen I am working on this, my current idea Is:
- Since we have a range as input parameter instead of a position, I am adding a new query method to the index that will return all the entities sitting in that range
- map all the entities to their code hints and return a flatten list of them

What do you think?

**Comment by magnus-madsen** [comment]
@nicdard I am not sure what the "right" solution is yet. I think the easiest is to look up all entities in the given uri. Then you filter and collect only the "LocalVar" onces. Next you translate each into a InlayHint with its type. This should correspond approximately to adding hints for local variable bindings. I would start there and we can see how it works.

**Comment by nicdard** [comment]
@magnus-madsen I have pushed an initial version to show the LocalVar implementation (still needs to be completed)

I believe however that we should use the range parameter from the request to filter out the entities (as specified in the protocol) instead of just computing the hints for the whole file, one commit implements the functionality in the index.

**Comment by magnus-madsen** [comment]
> @magnus-madsen I have pushed an initial version to show the LocalVar implementation (still needs to be completed)
> 
> I believe however that we should use the range parameter from the request to filter out the entities (as specified in the protocol) instead of just computing the hints for the whole file, one commit implements the functionality in the index.

Overall approach looks reasonable. We probably want a few helper functions here and there, but otherwise I think it makes sense 👍

File: main/src/ca/uwaterloo/flix/api/lsp/Index.scala [hunk 1]
```diff
@@ -180,6 +180,20 @@ case class Index(m: Map[(String, Int), List[Entity]],
...
 
+  /**
+    * Returns the expressions in the document at the given `uri` in the given range `range`.
+    */
+  def queryByRange(uri: String, range: Range): List[Entity] = {
+    (range.start.line to range.end.line).flatMap { line =>
+      m.getOrElse((uri, line), Nil)
+    }.filter { c =>
+      // Exclude all entities which are bigger than the given range.
+      c.loc.endLine <= range.end.line && c.loc.endCol <= range.end.character &&
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/api/lsp/Index.scala:191]: I think a helper function here would be useful.

File: main/src/ca/uwaterloo/flix/api/lsp/provider/InlayHintProvider.scala [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+  }
+
+  private def getTypeHint(loc: SourceLocation, tpe: Type): Option[InlayHint] = {
+    val pos = Position(loc.endLine, loc.endCol)
+    getTypeText(tpe).map(label =>
+      InlayHint(pos, label, Some(InlayHintKind.Type), None, "", paddingRight = true)
+    )
+  }
+
+  private def getTypeText(tpe: Type): Option[String] = {
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/api/lsp/provider/InlayHintProvider.scala:44]: We actually have a very nice class for this `FormatType.formatWellKindedType`.

File: main/src/ca/uwaterloo/flix/api/lsp/provider/InlayHintProvider.scala [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+        case (acc, _) =>
+          acc
+      }
+    }
+  }
+
+  private def getTypeHint(loc: SourceLocation, tpe: Type): Option[InlayHint] = {
+    val pos = Position(loc.endLine, loc.endCol)
+    getTypeText(tpe).map(label =>
+      InlayHint(pos, label, Some(InlayHintKind.Type), None, "", paddingRight = true)
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/api/lsp/provider/InlayHintProvider.scala:40]: I guess we have to wait for actual VSCode support to see how this should be formatted.

File: main/src/ca/uwaterloo/flix/api/lsp/provider/InlayHintProvider.scala (hunk 2 above)
Comment (nicdard) [reply main/src/ca/uwaterloo/flix/api/lsp/provider/InlayHintProvider.scala:44]: Oh that's nice, I didn't know!

File: main/src/ca/uwaterloo/flix/api/lsp/provider/InlayHintProvider.scala (hunk 3 above)
Comment (nicdard) [reply main/src/ca/uwaterloo/flix/api/lsp/provider/InlayHintProvider.scala:40]: Yes, for now it's just my conjecture from what I have seen in other projects

File: main/src/ca/uwaterloo/flix/api/lsp/Index.scala (hunk 1 above)
Comment (nicdard) [reply main/src/ca/uwaterloo/flix/api/lsp/Index.scala:191]: ok!

**Comment by nicdard** [comment]
Inlay hint as been finalized! I will start to take a look soon at the vscode part

**Comment by nicdard** [comment]
> Overall looks solid. I have some minor comments about documentation, but I think those can wait.
> 
> Once this is integrated with VSCode, I will do a more detailed pass.
> 
> I guess we have to wait for LSP 3.17 to be stable or?

Yes you are right, sorry, I was just looking that they finalized it but the new version of LSP is still not out. However I will try to implement the support in the extension usig vscode api for now, as it seems that the protocol is updated not so frequently

**Comment by magnus-madsen** [comment]
LSP 3.17 has been released. Thus I think it should be possible to start testing this PR. (with a suitable extension of the plugin too).

**Comment by nicdard** [comment]
> LSP 3.17 has been released. Thus I think it should be possible to start testing this PR. (with a suitable extension of the plugin too).

Very cool! :) I am going to start working back on this probably the 11th since I have 3 exams in the upcoming days :sweat:

**Comment by nicdard** [comment]
> > > LSP 3.17 has been released. Thus I think it should be possible to start testing this PR. (with a suitable extension of the plugin too).
> > 
> > 
> > Very cool! :) I am going to start working back on this probably the 11th since I have 3 exams in the upcoming days sweat
> 
> Good luck with the exams :)

Thanks!!

**Comment by magnus-madsen** [comment]
Is this ready for review or do you have some changes you want to make first?

**Comment by nicdard** [comment]
> Is this ready for review or do you have some changes you want to make first?

I still need to add some changes 💪

**Comment by nicdard** [comment]
It now works, but i still need to fix the position where they are displayed.

Screenshot from 2022-07-03 02-15-12

**Comment by nicdard** [comment]
I have adjusted the alignment :muscle:
Screenshot from 2022-07-03 02-30-08

**Comment by magnus-madsen** [comment]
!!!jar

File: main/src/ca/uwaterloo/flix/api/lsp/Entity.scala [hunk 4]
```diff
@@ -20,6 +20,11 @@ import ca.uwaterloo.flix.language.ast.{Name, SourceLocation, Symbol, TypedAst}
 
 sealed trait Entity {
   def loc: SourceLocation
+
+  def isInRange(range: Range): Boolean = {
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/api/lsp/Entity.scala:24]: Could you I ask you to add a comment? Does it mean that `range` is contained with entity or part of range is or what? :)

**Comment by magnus-madsen** [comment]
How does it work on these code fragments:

```
let x: Int32 = 123
```

```
let f = x -> x + 1;
f(123)
```

```
let _ = #{
  A(123).
  B(x) :- A(x).
```

Thanks! :)
```

File: main/src/ca/uwaterloo/flix/api/lsp/Entity.scala (hunk 4 above)
Comment (nicdard) [reply main/src/ca/uwaterloo/flix/api/lsp/Entity.scala:28]: This calculates wether the entity is fully included in a range, which can be a multiline range.

Comment (magnus-madsen) [reply main/src/ca/uwaterloo/flix/api/lsp/Entity.scala:28]: OK, can you you try to add comment? :)

Comment (nicdard) [reply main/src/ca/uwaterloo/flix/api/lsp/Entity.scala:28]: should be there already :)

**Comment by nicdard** [comment]
> How does it work on these code fragments:
> 
> ```
> let x: Int32 = 123
> ```
> 
> ```
> let f = x -> x + 1;
> f(123)
> ```
> 
> ```
> let _ = #{
>   A(123).
>   B(x) :- A(x).
> ```
> 
> Thanks!

So I need not to show the hint when the type is already there, for the function the type is correctly displayed, the last one example, I think a } is missing right?

**Comment by magnus-madsen** [comment]
> > How does it work on these code fragments:
> > ```
> > let x: Int32 = 123
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > ```
> > let f = x -> x + 1;
> > f(123)
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > ```
> > let _ = #{
> >   A(123).
> >   B(x) :- A(x).
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Thanks!
> 
> So I need not to show the hint when the type is already there, for the function the type is correctly displayed, the last one example, I think a } is missing right?

I am asking :) I was thinking you could paste a screen shot

File: main/src/ca/uwaterloo/flix/api/lsp/Entity.scala (hunk 4 above)
Comment (magnus-madsen) [reply main/src/ca/uwaterloo/flix/api/lsp/Entity.scala:28]: I meant something like:

```
/**
  * Returns `true` if the given range `range` is fully included in `this` entity (i.e. the given range must starter later and end earlier.)
  */
```

Comment (nicdard) [reply main/src/ca/uwaterloo/flix/api/lsp/Entity.scala:28]: ok!

**Comment by nicdard** [comment]
> > > How does it work on these code fragments:
> > > ```
> > > let x: Int32 = 123
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > ```
> > > let f = x -> x + 1;
> > > f(123)
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > ```
> > > let _ = #{
> > >   A(123).
> > >   B(x) :- A(x).
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > Thanks!
> > 
> > 
> > So I need not to show the hint when the type is already there, for the function the type is correctly displayed, the last one example, I think a } is missing right?
> 
> I am asking :) I was thinking you could paste a screen shot

Sure thing:

Screenshot from 2022-07-03 22-58-08



Screenshot from 2022-07-03 22-57-49


> > > How does it work on these code fragments:
> > > ```
> > > let x: Int32 = 123
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > ```
> > > let f = x -> x + 1;
> > > f(123)
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > ```
> > > let _ = #{
> > >   A(123).
> > >   B(x) :- A(x).
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > Thanks!
> > 
> > 
> > So I need not to show the hint when the type is already there, for the function the type is correctly displayed, the last one example, I think a } is missing right?
> 
> I am asking :) I was thinking you could paste a screen shot

**Comment by nicdard** [comment]
!!!jar

**Comment by magnus-madsen** [comment]
Thanks!

**Comment by nicdard** [comment]
Magnus sorry I am still working on it
I need to try out one thing, so that we don't get that double type when the type is given explicitly

**Comment by magnus-madsen** [comment]
Ah- I merged it, but you can just open a new PR. There are many improvements we can still make :)

**Comment by magnus-madsen** [comment]
For example, sometimes the types get very long. Maybe then we should hide them?

**Comment by nicdard** [comment]
> Ah- I merged it, but you can just open a new PR. There are many improvements we can still make :)

Yeah sure :)

**Comment by nicdard** [comment]
> For example, sometimes the types get very long. Maybe then we should hide them?

I think for this there should be an option for the label, I need to check

**Pull request merged**